src/crimson/osd/recovery_backend.cc

```diff
@@ -42,3 +45,227 @@ void RecoveryBackend::clean_up(ceph::os::Transaction& t,
   recovering.clear();
 }
 
+void RecoveryBackend::handle_backfill_finish(
+  MOSDPGBackfill& m)
+{
+  logger().debug("{}",
+                 __func__);
+  ceph_assert(crimson::common::local_conf()->osd_kill_backfill_at != 1);
+  auto reply = make_message<MOSDPGBackfill>(
+    MOSDPGBackfill::OP_BACKFILL_FINISH_ACK,
+    pg.get_osdmap_epoch(),
+    m.query_epoch,
+    spg_t(pg.get_pgid().pgid, pg.get_primary().shard));
+  reply->set_priority(pg.get_recovery_op_priority());
+  (void)m.get_connection()->send(std::move(reply));
+  // XXX: whh not PeeringState::RecoveryDone?
```

Comment:
Probably because you don't want to hold up the message processing on the async peering state event.